### ISSUE TITLE
How could SPANN load DEEP-1B (358GB) into memory with only 128GB RAM (in SPANN paper)?

### ISSUE THREAD

**Issue by symphony233** [microsoft/SPTAG#268]
Hi, I'm reading the source code and paper of SPANN.
In `/AnnService/src/IndexBuilder/main.cpp`, we can see data will be loaded in function `DefaultVectorReader::GetVectorSet()`.
In that function, vectors are loaded in one single `ReadBinary()` method.

But if the size of origin file exceeds the RAM, how could do that? In SPANN paper, only uses 128GB RAM for DEEP-1B, which has 358GB basepoints file according to [big-ann-benchmark](https://big-ann-benchmarks.com/).

**Comment by LLLjun** [comment]
Hi, have you solved this problem? I also encountered the same.

**Comment by matchyc** [comment]
No, I didn't. And I suppose the team use more footprint of memory than they described in paper? That does not mean they were wrong, the description of memory usage in paper is all about searching procedure, so maybe in the building step, spann needs more memory space. I guess...



| |
***@***.***
|
|
***@***.***
|




---- Replied Message ----
| From | ***@***.***> |
| Date | 08/12/2022 16:22 |
| To | ***@***.***> |
| Cc | ***@***.******@***.***> |
| Subject | Re: [microsoft/SPTAG] How could SPANN load DEEP-1B (358GB) into memory with only 128GB RAM (in SPANN paper)? (Issue #268) |

Hi, have you solved this problem? I also encountered the same.

**Comment by matchyc** [comment]
> Hi, have you solved this problem? I also encountered the same.

No, I didn't. And I suppose the team uses more memory footprint than they described in paper? That does not mean they were wrong, the description of memory usage in the paper is all about searching procedure, so maybe in the building step, spann needs more memory space. I guess...

**Comment by LLLjun** [comment]
Got it, thanks.

**Comment by LLLjun** [comment]
By the way, how much memory does it take to build the SPANN index using the DEEP1B dataset?

**Comment by matchyc** [comment]
No idea, if you figure out I'd like to hear it! Thanks ahead.

**Comment by matchyc** [comment]
> By the way, how much memory does it take to build the SPANN index using the DEEP1B dataset?

No idea, if you figure out I'd like to hear it! Thanks ahead.

**Comment by laa** [comment]
Guys, that is only a partial implementation of SPAN. It only implements a fast search of centroids using trees. The rest you need to do itself.

**Comment by lolz0r** [comment]
Thanks for writing this up; it is helpful to know that memory savings only happen in search (so if you have a really big DB, you need tons of memory to build the database)

**Comment by suppersam1** [comment]
maybe need to use the BalancedDataPartition program to perform data partitioning and then build it on each machine in a distributed manner.
